auto_uids.h

```diff
@@ -1,16 +1,29 @@
 #ifndef AUTO_UIDS_H
 #define AUTO_UIDS_H
 
-extern int auto_uida;
-extern int auto_uidd;
-extern int auto_uidl;
-extern int auto_uido;
-extern int auto_uidp;
-extern int auto_uidq;
-extern int auto_uidr;
-extern int auto_uids;
```

Comment:
Temperature check for the team.  These variables can be set at runtime via running a procedure using the init section of an ELF executable.  Example:

    #include <sys/types.h>
    #include <pwd.h>
    #include <stdlib.h>
    #include <stdio.h>

    static uid_t root_uid = -1;
    static void lookup(void) __attribute__((constructor));

    static void
    lookup(void)
    {
        struct passwd *pw;
        pw = getpwnam("root");
        if(!pw) { exit(1); }
        root_uid = pw->pw_uid;
    }

    int
    main(int argc, char *argv[])
    {
        printf("root uid: %ld\n", (long)root_uid);
        exit(0);
    }

which produces the output:

    root uid: 0

Here the procedure lookup, due to the `__attribute__((constructor))` directive, is added to the .init section of the executable and run before main.  This is the mechanism that C++ uses for static initialization.

If we used this feature we'd be able to retain the `auto_` variables here, deferring when they are populated to runtime.  Further we could use different values in different executable programs, so this same feature could be used for the fake_chown work this patch does.

Overall I think using .init to populate these variables would lead to a simpler patch, at the expense of introducing a new compiler construct to the code.  Said construct is widely available, owing to C++ requiring it, so I believe the cost to be modest and worth using here, but I'm interested in what everyone thinks about this alternative.

Reply:
This is cool! TIL. My main worry is portability. What happens on non-ELF systems like OS X? How recent a compiler would we be requiring?

Reply:
Here is the same program on OS X, courtesy of hmadison.  This construct works on OS X, producing the same output I report above on Linux.

https://gist.github.com/hmadison/4ecab14ddbcf3bd651ab848e6d578d53

The test included in it's entirety here:

    hmadison@hunters-mbp ~/s/test> cat alynpost.c 
    #include <sys/types.h>
    #include <pwd.h>
    #include <stdlib.h>
    #include <stdio.h>

    static uid_t root_uid = -1;
    static void lookup(void) __attribute__((constructor));

    static void
    lookup(void)
    {
        struct passwd *pw;
        pw = getpwnam("root");
        if(!pw) { exit(1); }
        root_uid = pw->pw_uid;
    }

    int
    main(int argc, char *argv[])
    {
        printf("root uid: %ld\n", (long)root_uid);
        exit(0);
    }
    hmadison@hunters-mbp ~/s/test> gcc alynpost.c 
    hmadison@hunters-mbp ~/s/test> ./a.out 
    root uid: 0
    hmadison@hunters-mbp ~/s/test> 
    hmadison@hunters-mbp ~/s/test> uname -a
    Darwin hunters-mbp.lan 18.6.0 Darwin Kernel Version 18.6.0: Thu Apr 25 23:16:27 PDT 2019; root:xnu-4903.261.4~2/RELEASE_X86_64 x86_64

Reply:
This sounds useful, but this should be discussed in it's own issue or PR, or maybe on some mailing list.

Reply:
BTW, I can't remember why I bundled "looking up IDs at runtime" and "staged installs" into a single patch -- maybe because deferring IDs at runtime let me do staged installs as a non-root user? -- but if we can unbundle the two features and evaluate implementations of each on its own merits, we probably should.

Reply:
> This sounds useful, but this should be discussed in it's own issue or PR, or maybe on some mailing list.

Fully agreed.  I've submitted PR#11 that covers the runtime lookup of uid/gid in the manner I describe here.